cloud_storage/segment_meta_cstore: support for insert in the middle via a write_buffer

andijcr opened the pull request:
segment_meta_cstore now support inserting entries that replaces previously inserted one. these new entries have to be [base_offset, committed_offset] aligned to a group of pre-existing entries.
insertion is performed by cloning the old frames up to the frame containing entries that will be replaced. after that, the rest of the frames are rebuilt by merging in the new segment_meta and discarding the one replaced. 
To speed up this operation, a write_buffer is used to buffer operations so as to coalesce various operations into one rebuild.

Fixes #9422 
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md##pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.1.x
- [ ] v22.3.x
- [ ] v22.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
* none

Lazin commented on src/v/cloud_storage/segment_meta_cstore.cc:
```diff
@@ -563,65 +573,135 @@ class segment_meta_cstore::impl
+
+        auto append_entries_begin = partition_point;
+        auto append_entries_end = write_buffer.end();
+        // TODO bad_alloc safety?
```
The `append` itself is transactional but here I think you need to make transaction span multiple elements, so modified version of the `append` is needed.

andijcr commented on src/v/cloud_storage/segment_meta_cstore.cc:
```diff
@@ -230,6 +333,89 @@ class column_store
+              = get_baseoffset_frame_for_offset(*offset_seg_it);
+            // find next offset_seg_it that resides in a different frame
+            auto next_offset_seg_it = std::find_if_not(
+              ++offset_seg_it,
```
it's hiding a bug, good catch. this part is being rewritten to address the edge cases

dotnwat commented on src/v/cloud_storage/segment_meta_cstore.cc:
```diff
@@ -86,13 +86,42 @@ void increment_all(std::tuple<Args...>& tup) {
 
+/// zip first_tuple with second_tuple, calls map_fn on each pair of elements,
+/// collect results in a tuple if map_fn produces a result
+template<typename Fn>
+auto tuple_map(Fn&& map_fn, auto&& first_tuple, auto&& second_tuple) {
```
you might be able to use boost hana or some ranges stuff for this.

andijcr replied:
waiting for clang15 to be able to use ranges, and boost hana i didn't consider it

dotnwat commented on src/v/cloud_storage/segment_meta_cstore.cc:
```diff
@@ -230,6 +296,129 @@ class column_store
         }
     }
 
+    auto serde_fields() { return std::tuple_cat(columns(), std::tie(_hints)); }
```
1. does this now create a copy?
2. it seems error prone: with serde_fields we basically only get away with appends. but with this formulation it would seem we could only add hints without needing to abandon serde fields mechanism?

dotnwat replied:
>could you elaborate on this point?

what is the intended way to extend the data structure? in this formulation you could only append to `_hints` with a basic version bump

andijcr replied:
if i understand the question correctly: yes like in  other cases, new fields would have to be added ` ... std::tie(_hints, _new_field)`
But probably if there is a need to change column_store, chances are that it will be done in serde_write/read. 

...So probably it's another argument to ditch serde_fields here and use another name

dotnwat commented on src/v/cloud_storage/segment_meta_cstore.cc:
```diff
@@ -563,65 +743,159 @@ class segment_meta_cstore::impl
 
-    auto serde_fields() { return std::tie(_col); }
+    auto serde_fields() {
+        flush_write_buffer();
```
this is hard to understand. i think the caller should do this?

dotnwat commented on src/v/cloud_storage/segment_meta_cstore.cc:
```diff
@@ -563,65 +743,159 @@ class segment_meta_cstore::impl
+        _write_buffer.clear();
+    }
+
+    mutable absl::btree_map<model::offset, segment_meta> _write_buffer{};
+    mutable column_store _col{};
```
its very rare to see mutable members. why do we need this?

andijcr replied:
It's to be able to maintain the const qualifiers on members. Logically the state of the object does not change, but I need to flush to be able to call begin().
Truth to be told, I added flush to be able to get something working correctly to test; there is a plan to expand the iterator to perform the merging on the fly, so that and some other improvements would enable to remove mutable

dotnwat replied:
thanks for the explanation

dotnwat reviewed (commented):
do exists tests already exercise all the new code?

andijcr commented on src/v/cloud_storage/segment_meta_cstore.cc:
```diff
@@ -563,65 +573,135 @@ class segment_meta_cstore::impl
 
-    size_t size() const { return _col.size(); }
+    size_t size() const {
+        // TODO how to deal with write_buffer?
```
I'm still on the fence about this. Since the result could be used for at_index, it needs to be exact. On the other hand, at_index does not look like a core operation that needs to be there, so it could be moved to a private context, and size could be renamed to "approximate_size"

andijcr commented:
> do exists tests already exercise all the new code?

Test in the last commit, but I'm looking at ways to simplify it a bit

Lazin commented:
How the read path is going to look like?

dotnwat commented on src/v/cloud_storage/segment_meta_cstore.cc:
```diff
@@ -41,7 +41,7 @@ static constexpr uint32_t sampling_rate = 8;
   gauge_col_t::max_frame_size % sampling_rate == 0, "Invalid sampling rate");
 
-enum class segment_meta_ix {
+enum class segment_meta_ix : size_t {
```
why specify an underlying type here?

andijcr replied:
It's temp code. I didn't pay attention to this enum until the last commit, where I understood that the enum is used as a named index into the tuple-of-iterators. I specified the type as a reminder for myself.

andijcr replied:
btw I'm moving the code to another pr https://github.com/redpanda-data/redpanda/pull/9674 so this one can be reviewed more easily while I work on that one

Lazin commented on src/v/cloud_storage/segment_meta_cstore.cc:
```diff
@@ -563,65 +794,159 @@ class segment_meta_cstore::impl
+        // (different base_offset), but still overlaps with the successors? i.e.
+        // a gap in the data where a previous segment is lost, and the current
+        // one would be a replacement for the previous one?
+        if(_write_buffer.size()>1 &&
```
This logic is completed, I guess that we just need to replace an element, the replacement logic in the manifest is handled explicitly, there is a method that first finds all elements that will be replaced and moves them to the `partition_manifest::_replaced` collection. And this component should probably behave just like a `btree_map`.

Lazin replied:
Also, maybe there is no real need to return a fragmented_vector of replaced segments from the `replace_entries` for the same reason.

Lazin reviewed (commented):
In general, this looks good. Some nits.

Lazin commented on src/v/cloud_storage/segment_meta_cstore.cc:
```diff
@@ -563,65 +794,159 @@ class segment_meta_cstore::impl
 
-    void insert(const segment_meta& m) { _col.append(m); }
+    void insert(const segment_meta& m) {
+        auto [m_it, insertion_took_place] = _write_buffer.insert_or_assign(
```
Also, maybe the logic will get a bit more straightforward if you will check for the overlap before insertion. Then, some action could be taken on replaced element inside the `_write_buffer` instead of figuring out what was replaced afterwards.

andijcr replied:
so in the end column_store::insert_entries handles all the buffer directly. this makes the append-only operation still fast, since the whole merging section is skipped. It also buys atomicity, since it's done on list of clones that will be swapped at the end of operation.

andijcr commented on src/v/cloud_storage/segment_meta_cstore.cc:
```diff
@@ -563,65 +794,159 @@ class segment_meta_cstore::impl
-    auto inflated_actual_size() const { return _col.inflated_actual_size(); }
+    auto inflated_actual_size() const {
+        // TODO how to deal with write_buffer? is it ok to return an approx
+        // value by not flushing?
```
This holds true only if we keep a vector of replaced segments as a member, right? otherwise, without performing the merge operation, it's not possible to know exactly this value (not that this needs to be exact, i think)

andijcr commented on src/v/cloud_storage/segment_meta_cstore.cc:
```diff
@@ -563,65 +794,159 @@ class segment_meta_cstore::impl
     void append(const segment_meta& meta) { _col.append(meta); }
 
     std::unique_ptr<segment_meta_materializing_iterator::impl> begin() const {
+        flush_write_buffer();
```
Separated in the next pr, keeping this here to keep it working separately.

andijcr commented on src/v/cloud_storage/segment_meta_cstore.cc:
```diff
@@ -230,6 +299,133 @@ class column_store
+        while (old_segments_it != old_segments_end
+               && offset_seg_it != offset_seg_end) {
+            auto [replacement_base_offset, replacement_meta] = *offset_seg_it;
+            ++offset_seg_it;
```
I think I can turn it more easily with the next pr, where I have a proper merge function to reuse instead of this while

andijcr commented on src/v/cloud_storage/segment_meta_cstore.cc:
```diff
@@ -230,6 +299,133 @@ class column_store
+     */
+    template<std::input_iterator IT>
+    auto
+    replace_entries(IT offset_seg_it, std::sentinel_for<IT> auto offset_seg_end)
```
Fixed, it was from when the types were still fuzzy

andijcr commented on src/v/cloud_storage/segment_meta_cstore.cc:
```diff
@@ -563,65 +794,159 @@ class segment_meta_cstore::impl
+              model::offset::min());
+            auto buffer_last_base_offset = _write_buffer.crbegin()->first;
+            // CHECK: base_offset is never equal to committed_offset
+            if (likely(buffer_last_base_offset > last_committed_offset)) {
```
I modified the code; now it uses committed_offset to choose. I modified the test also to check this function

Lazin commented on src/v/cloud_storage/segment_meta_cstore.cc:
```diff
@@ -154,6 +183,53 @@ class column_store
+      [](segment_meta const& s) { return s.ntp_revision(); },
+      [](segment_meta const& s) { return s.archiver_term(); },
+      [](segment_meta const& s) { return s.segment_term(); },
+      [](segment_meta const& s) { return s.delta_offset_end(); },
```
oh, this PR is not merged yet

andijcr replied:
Ah right, the field introduced by @andrwng ?

andijcr commented:
force push: rebased on dev, added static assert over number of fields of segment_meta, found a seed for failing unit test:
`cloud_storage_basic_rpunit -t test_segment_meta_cstore_prefix_truncate_full -- -c 1`
seed that will cause a failure "10263162". uncomment this line to try it https://github.com/redpanda-data/redpanda/pull/9420/files#diff-d6f3d11f44e0ccc9114109c058d1561b8ffa9af1229ba0db10dcdb71ae8c5922R749

andijcr commented on src/v/cloud_storage/segment_meta_cstore.cc:
```diff
@@ -370,9 +573,17 @@ class column_store
         return materialize(std::move(it));
     }
 
+    void clear() {
+        // replace *this with an empty instance
+        auto replacement = column_store{};
+        auto current_fields = member_fields();
+        auto new_fields = replacement.member_fields();
+        std::swap(current_fields, new_fields);
+    }
     void prefix_truncate(int64_t bo) {
         auto lb = _base_offset.lower_bound(bo);
         if (lb == _base_offset.end()) {
+            clear();
             return;
         }
```
@Lazin clear seems missing here; it was the reason test_segment_meta_cstore_prefix_truncate_full was failing for some inputs. It's probably a branch that would never be taken if _write_buffer wasn't there
